src/block_service/directio_writable_file.cpp

```diff
@@ -0,0 +1,137 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <algorithm>
+#include <cstring>
+#include <fcntl.h>
+#include <stdlib.h> // posix_memalign
+#include <string>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h> // getpagesize
+
+#include <dsn/dist/fmt_logging.h>
+#include <dsn/utility/flags.h>
+
+#include "block_service/directio_writable_file.h"
+
+DSN_DEFINE_uint32("replication",
+                  direct_io_buffer_pages,
+                  64,
+                  "Number of pages we need to set to direct io buffer");
+DSN_TAG_VARIABLE(direct_io_buffer_pages, FT_MUTABLE);
+
+DSN_DEFINE_bool("replication",
+                enable_direct_io,
+                false,
+                "Whether to enable direct I/O when download files");
+DSN_TAG_VARIABLE(enable_direct_io, FT_MUTABLE);
+
+namespace dsn {
+namespace dist {
+namespace block_service {
+
+const uint32_t g_page_size = getpagesize();
+
+direct_io_writable_file::direct_io_writable_file(const std::string &file_path)
+    : _file_path(file_path),
+      _fd(-1),
+      _file_size(0),
+      _buffer(),
+      _buffer_size(FLAGS_direct_io_buffer_pages * g_page_size),
+      _offset(0)
+{
+}
+
+direct_io_writable_file::~direct_io_writable_file()
+{
+    if (!_buffer || _fd < 0) {
+        return;
+    }
+    // Here is an ensurance, users shuold call finalize manually
+    if (_offset > 0 && !finalize()) {
+        derror_f("Call finalize return error in destructor");
+    }
+    free(_buffer);
+    close(_fd);
+}
+
+bool direct_io_writable_file::initialize()
+{
+    if (posix_memalign(&_buffer, g_page_size, _buffer_size) != 0) {
+        derror_f("Allocate memaligned buffer failed, errno = {}", errno);
+        return false;
+    }
+
+    int flag = O_WRONLY | O_TRUNC | O_CREAT | O_DIRECT;
+    _fd = open(_file_path.c_str(), flag, S_IRUSR | S_IWUSR | S_IRGRP);
+    if (_fd < 0) {
+        derror_f("Failed to open {} with flag {}, errno = {}", _file_path, flag, errno);
+        free(_buffer);
+        _buffer = nullptr;
+        return false;
+    }
+    return true;
+}
+
+bool direct_io_writable_file::finalize()
+{
+    if (!_buffer || _fd < 0) {
+        derror_f("Initialize the instance first");
+        return false;
+    }
+    if (_offset > 0) {
+        if (::write(_fd, _buffer, _buffer_size) != _buffer_size) {
+            derror_f("Failed to write last chunk, filie_path = {}, errno = {}", _file_path, errno);
+            return false;
+        }
+        _offset = 0;
+        ftruncate(_fd, _file_size);
+    }
+    return true;
+}
+
+uint32_t direct_io_writable_file::write(const char *s, uint32_t n)
+{
+    if (!_buffer || _fd < 0) {
+        derror_f("initialize the instance first");
+        return -1;
```

Comment:
`-1` is not an `uint32_t`.
Any more, I think use asserts would be better than defensive programing, all the code is controlled by programers, this case will not happend if we use `direct_io_writable_file` correctly, i.e. calling sequence is ordered in ctor() -> initialize() -> write() -> finalize() -> dector().

Reply:
Thinks.